### PR TITLE
build: add cmake parameters to determine msgpack to json buffer strategy.

### PR DESCRIPTION
**Pull request by franciscovalentecastro**
Signed-off-by: Francisco Valente <fcovalente@google.com>

Adding build parameters to determine the buffer strategy used in the function `flb_msgpack_raw_to_json_sds`. This is a way to solve [#5156](https://github.com/fluent/fluent-bit/issues/5156) by enabling the user to set parameters closer to what was set before https://github.com/fluent/fluent-bit/commit/b884764a386be8b327aa9ee9288ae1490ac116d1 .

**Explanation**

The parameter `FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE` is a multiplier to determine the initial buffer size to store the conversion to json from a msgpack object. It multiplies the size of the buffer where the msgpack object is stored to obtain an estimate of the allocation required. e.g. A value of 2 sets to allocate a buffer of size 2x of the received payload.

The parameter `FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE` is a multiplier to determine the size in bytes to add in every memory reallocation when more space is needed for the buffer. It multiplies the size of the buffer where the msgpack object is stored to obtain a percentage of memory to be increased each round. e.g. A value of .1 sets to increase in rounds of 10% of the original payload size.

This parameters can be set during build configuration in the following way :
 
``` 
cmake ../ -DFLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE=1.5 \
   -DFLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE=.10
```

**Valgrind**
Executed with sample configuration : 

```
[INPUT]
    Buffer_Chunk_Size 512k
    Buffer_Max_Size   5M
    Key               message
    Mem_Buf_Limit     10M
    Name              tail
    Path              /var/log/syslog,/var/log/message
    Read_from_Head    True
    Rotate_Wait       30
    Skip_Long_Lines   On
    Tag               default_pipeline.syslog

[OUTPUT]
    Match_Regex                   ^(default_pipeline\.syslog)$
    Name                          stackdriver
    Retry_Limit                   3
    net.connect_timeout_log_error False
    resource                      gce_instance
    stackdriver_agent             Google-Cloud-Ops-Agent-Logging/2.11.0 (BuildDistro=buster;Platform=linux;ShortName=debian;ShortVersion=10.11)
    tls                           On
    tls.verify                    Off
    workers                       8
```
Valgrind execution : 
```
valgrind fluent-bit/build/bin/fluent-bit   --config ./fluent_bit_main.conf
==7694== Memcheck, a memory error detector
==7694== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==7694== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==7694== Command: fluent-bit/build/bin/fluent-bit --config ./fluent_bit_main.conf
==7694== 
Fluent Bit v1.9.2
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/04/07 21:19:03] [ info] [fluent bit] version=1.9.2, commit=404db2aaad, pid=7694
[2022/04/07 21:19:03] [ info] [storage] version=1.1.6, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/04/07 21:19:03] [ info] [cmetrics] version=0.3.0
[2022/04/07 21:19:05] [ info] [output:stackdriver:stackdriver.0] metadata_server set to http://metadata.google.internal
[2022/04/07 21:19:05] [ warn] [output:stackdriver:stackdriver.0] client_email is not defined, using a default one
[2022/04/07 21:19:05] [ warn] [output:stackdriver:stackdriver.0] private_key is not defined, fetching it from metadata server
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #0 started
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #1 started
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #2 started
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #3 started
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #4 started
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #5 started
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #6 started
[2022/04/07 21:19:07] [ info] [output:stackdriver:stackdriver.0] worker #7 started
[2022/04/07 21:19:07] [ info] [sp] stream processor started
[2022/04/07 21:19:07] [ info] [input:tail:tail.0] inotify_fs_add(): inode=407396 watch_fd=1 name=/var/log/syslog
^C[2022/04/07 21:19:36] [engine] caught signal (SIGINT)
[2022/04/07 21:19:36] [ info] [input] pausing tail.0
[2022/04/07 21:19:36] [ warn] [engine] service will shutdown in max 5 seconds
[2022/04/07 21:19:37] [ info] [engine] service has stopped (0 pending tasks)
[2022/04/07 21:19:37] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=407396 watch_fd=1
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #0 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #0 stopped
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #1 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #1 stopped
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #2 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #2 stopped
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #3 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #3 stopped
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #4 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #4 stopped
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #5 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #5 stopped
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #6 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #6 stopped
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #7 stopping...
[2022/04/07 21:19:37] [ info] [output:stackdriver:stackdriver.0] thread worker #7 stopped
==7694== 
==7694== HEAP SUMMARY:
==7694==     in use at exit: 0 bytes in 0 blocks
==7694==   total heap usage: 63,773 allocs, 63,773 frees, 59,970,423 bytes allocated
==7694== 
==7694== All heap blocks were freed -- no leaks are possible
==7694== 
==7694== For counts of detected and suppressed errors, rerun with: -v
==7694== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [x] Documentation required for this feature
Adding documentation for new build parameters under new section `Build Options \ Optimization Options` : https://github.com/fluent/fluent-bit-docs/pull/786

**Backporting**

- [x] Backport to latest stable release.
Backport to `1.8` in https://github.com/franciscovalentecastro/fluent-bit/commit/b92e8bb3b66e926cd05cd16e2527eb3f4b725aa5.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by lecaros** [comment]
Hey @franciscovalentecastro IRT docs, these params should be under the Optimization section; it should be created at the same level as General, Development, etc.

**Comment by franciscovalentecastro** [comment]
Hi @lecaros . Created a documentation PR https://github.com/fluent/fluent-bit-docs/pull/786 with the parameters under a new section (`Build Options \ Optimization Options`) as you described.

**Comment by edsiper** [comment]
@franciscovalentecastro 

Have you checked the final values being configured when the macros are generated ? I have not tested but I am concerned the current variable definition uses a string instead of a number, the CMake macro expansion does this:

https://github.com/fluent/fluent-bit/blob/1fa0e94a09e4155f8a6d8a0efe36a5668cdc074e/cmake/macros.cmake#L8-L11

I suggest to add a simple unit test to check the defined and expected values

**Comment by franciscovalentecastro** [comment]
Hello @edsiper ! You're right, a simple unit test to check the defined and expected values should be a great addition. I'm going to start working on that.

I've only made simple verifications of the values with info messages like this :

```
flb_plg_info(ctx->ins, "FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE : %f", FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE);
flb_plg_info(ctx->ins, "FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE : %f", FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE);
```

```
[2022/04/19 20:16:24] [ info] [output:stackdriver:stackdriver.0] FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE : 1.500000
[2022/04/19 20:16:24] [ info] [output:stackdriver:stackdriver.0] FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE : 0.100000
```

**Comment by franciscovalentecastro** [comment]
Hello @edsiper ! You're right, a simple unit test to check the defined and expected values should be a great addition. I'm going to start working on that.

I've only made simple verifications of the values with info messages like this :

```
flb_plg_info(ctx->ins, "FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE : %f", FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE);
flb_plg_info(ctx->ins, "FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE : %f", FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE);
```

```
[2022/04/19 20:16:24] [ info] [output:stackdriver:stackdriver.0] FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE : 1.500000
[2022/04/19 20:16:24] [ info] [output:stackdriver:stackdriver.0] FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE : 0.100000
```

File: tests/internal/pack.c [hunk 1]
```diff
@@ -241,6 +241,14 @@ void test_json_pack_mult_iter()
...
 }
 
+/* Validate default values of macros used in flb_msgpack_raw_to_json_sds */
+void test_msgpack_to_json_macros()
+{
+    /* Verify default values */
+    TEST_CHECK(FLB_MSGPACK_TO_JSON_INIT_BUFFER_SIZE == 2.0);
+    TEST_CHECK(FLB_MSGPACK_TO_JSON_REALLOC_BUFFER_SIZE == 0.10);
+}
+
```
Comment (franciscovalentecastro) [inline tests/internal/pack.c:251]: @edsiper 

Added this unit test to verify the default values. I believe this also verifies the values are type `float` and not strings. Is there another way to test this without having to `hard code` the default values ?

**Comment by edsiper** [comment]
thanks!

**Pull request merged**